atoma-vllm/src/worker.rs

```diff
@@ -0,0 +1,690 @@
+                            "Empty prompts should not be received in `ModelWorker`".into(),
+                        ));
+                    }
+                    // DON'T PANIC: we should not receive empty prompts
+                    let last_token_id = sequence_data.get_last_token_id().unwrap();
+                    &[last_token_id]
+                };
+
+                // 4. Query length
+                let query_length = if is_prompt {
+                    sequence_length - context_length
+                } else {
+                    1
+                };
+
+                // 5. Update previous values if sliding window is used
+
+                // These are seq_len/context_len capped to the sliding window.
+                // They are passed to decode kernel.
+                // We still need original seq_len/context_len to compute slot
+                // mapping (and input position) below.
+                let mut sliding_sequence_length = sequence_length;
+                let sliding_context_length = context_length;
+
+                // This is a hack to make sliding window work with
+                // Paged Attention. We can remove it if we make paged attn kernel
+                // to properly handle sliding window attention.
+                if self.model.sliding_window().is_some() && !is_prompt {
+                    // DON'T PANIC: by the branch check
+                    sliding_sequence_length =
+                        self.model.sliding_window().unwrap().min(sequence_length);
+                }
+
+                // 6. Get block table for the current sequence
+                let block_table = if self.enable_chunked_prefill || !is_prompt {
+                    // DON'T PANIC: Unwrap is safe here because block_tables
+                    // should have allocated a block table for this sequence
+                    let mut block_table = sequence_group_metadata
+                        .block_tables
+                        .get(sequence_id)
+                        .expect("Block table should be allocated for sequence on decoding phase")
+                        .clone();
+
+                    // 7. If sliding window is used, we need to trim the block table
+                    if let Some(sliding_window) = self.model.sliding_window() {
+                        let sw_block_num = (sliding_window + self.cache_engine.block_size - 1)
+                            / self.cache_engine.block_size;
+                        let start = block_table.len().saturating_sub(sw_block_num);
+                        block_table = block_table[start..].to_vec();
+                    }
+
+                    block_table
+                } else {
+                    // Prefill without chunked prefill
+                    vec![]
+                };
+
+                // 8. Update intermediate states
+                block_tables.push(block_table);
+                sequence_lengths.push(sliding_sequence_length as f32);
```

Comment:
why are sequence_lengths f32?

Reply:
That's super annoying, it is necessary to handle `cumsum` on the defined tensor below.. as it seems that `matmul` is not defined for integer `dtype`s